### ISSUE TITLE
BUG: 1.13.2 release fails upon import (Mac OS X & Linux/Python 3.5.1)

### ISSUE THREAD

**Issue by adamosloizou** [numpy/numpy#9786]
#### Versions
- OS X: `10.12.1 (16B2555)`
- Linux: `Amazon Linux AMI release 2016.03`
- Python: `3.5.1`
- Pip: `9.0.1`
- Numpy: `1.13.2`
  - `numpy-1.13.2-cp35-cp35m-macosx_10_6_intel.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl` for Mac OS X
  - `numpy-1.13.2-cp35-cp35m-manylinux1_x86_64.whl` for Linux

#### Description
We are getting import errors when attempting to use `numpy`:

#### Mac OS X
```
Python 3.5.1 (default, May  9 2017, 17:14:37)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/adamos.loizou/.virtualenvs/tmp-f692dfaf032dece/lib/python3.5/site-packages/numpy/__init__.py", line 166, in <module>
    from . import random
  File "/Users/adamos.loizou/.virtualenvs/tmp-f692dfaf032dece/lib/python3.5/site-packages/numpy/random/__init__.py", line 99, in <module>
    from .mtrand import *
ImportError: dlopen(/Users/adamos.loizou/.virtualenvs/tmp-f692dfaf032dece/lib/python3.5/site-packages/numpy/random/mtrand.cpython-35m-darwin.so, 2): Symbol not found: __PyThreadState_UncheckedGet
  Referenced from: /Users/adamos.loizou/.virtualenvs/tmp-f692dfaf032dece/lib/python3.5/site-packages/numpy/random/mtrand.cpython-35m-darwin.so
  Expected in: flat namespace
 in /Users/adamos.loizou/.virtualenvs/tmp-f692dfaf032dece/lib/python3.5/site-packages/numpy/random/mtrand.cpython-35m-darwin.so
```

#### Linux
```
Python 3.5.1 (default, Sep 13 2016, 18:48:37)
[GCC 4.8.3 20140911 (Red Hat 4.8.3-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/venv/local/lib/python3.5/dist-packages/numpy/__init__.py", line 166, in <module>
    from . import random
  File "/root/venv/local/lib/python3.5/dist-packages/numpy/random/__init__.py", line 99, in <module>
    from .mtrand import *
ImportError: /root/venv/local/lib/python3.5/dist-packages/numpy/random/mtrand.cpython-35m-x86_64-linux-gnu.so: undefined symbol: _PyThreadState_UncheckedGet
```

This also **affects using pandas** upon a fresh installation because `pandas` only specifies a minimum `numpy` version thus taking the latest.

**Comment by charris** [comment]
Seems to be a cython error: https://github.com/cython/cython/issues/1880.

**Comment by charris** [comment]
I've hidden the 1.13.2 release on PyPi.

**Comment by matthew-brett** [comment]
I think that doesn't stop pip from seeing it....

**Comment by charris** [comment]
Pip will still see it, but it won't be the default download. Maybe I should just delete it ...

**Comment by pv** [comment]
Can you delete just the broken file (py3.5 OSX)?  IIRC it's not possible 
to re-upload a replacement file on Pypi, but maybe it's possible to 
delete it?

In any case, this looks like
(i) the Cython version has to be pinned to a known-good one in the 
numpy-wheels repository,
(ii) Numpy 1.13.3 should follow :(

**Comment by matthew-brett** [comment]
I think it will be the default still. Might be wrong.

**Comment by matthew-brett** [comment]
We can use a build tag to upload a fixed version for 3.5

On 28 Sep 2017 13:19, "Matthew Brett" <matthew.brett@gmail.com> wrote:

I think it will be the default still. Might be wrong.

**Comment by charris** [comment]
A build tag changes the file name?

**Comment by matthew-brett** [comment]
Yes it does . Sorry walking in London now otherwise would do it

**Comment by charris** [comment]
I removed the file, but I suspect PyPi will resort to the source. OTOH, who knows, it's PyPi ;)

**Comment by charris** [comment]
Hmm, this means the cythonized source files are also in error, so need to be uploaded also. @matthew-brett How does one add a build tag?

**Comment by charris** [comment]
Is it only 3.5 on OSX? Google seems to indicate that it is micro version dependent, so maybe just hasn't been noticed on other systems.

**Comment by pv** [comment]
The report above states also the manylinux 3.5 wheels are broken.

Re: build tags, https://github.com/MacPython/wiki/wiki/Build-Tags

**Comment by adamosloizou** [comment]
@charris please bear in mind that we are seeing failures in *both Mac OS and Linux* (although we get different error messages).
I don't know about Windows.

**Comment by charris** [comment]
OK, sounds as if all the 3.5 wheels need to be deleted.

**Comment by charris** [comment]
Looks like a build tag should be included in the appveyor and travis scripts.

**Comment by charris** [comment]
The discussion at #9787 might also be of interest. There is some indication that we may need to do another point release in order to support legacy scripts.

**Comment by carno** [comment]
Pulling down just the cp35 wheels does not really help people who install numpy in environments without a ready compilation environment (cloud for the win! ;-)), as pip tries to compile from source and fails due to lack of compiler and/or dev libs needed for compilation.

**Comment by charris** [comment]
@carno We can't push new wheels until the version of cython used in the build is pinned to something useable. That said, there is an argument to be made for simply removing the release from pip, as there is no easy way to mark it as broken, and fixing with build tags might break downstream scripts.

**Comment by jonparrott** [comment]
(drive by comment): it's generally discouraged to delete anything from PyPI.

**Comment by matthew-brett** [comment]
Chuck - can't we just rebuild with Cython 0.26?

I think we do have to remove the release, because we can't fix the source archive for 1.13.2.  How about 1.13.3 with the same content, but Cython 0.26 wheels / source?

**Comment by greatestape** [comment]
@charris I don't remember what the conventions are with pypi, but I think you should delete 1.13.2 and push 1.13.3 once the issue is resolved. As it is, anyone who hasn't pinned their numpy requirement (and even people who pinned to 1.13.*, which should be safe) is going to have their builds start failing (which is what happened to us)

**Comment by matthew-brett** [comment]
Just to clarify - this is all due to https://github.com/cython/cython/issues/1880

The bug appears when code generated with Cython 0.27 gets run on Python 3.5 < 3.5.2 - at least on Linux and OSX - maybe also on Windows.

**Comment by charris** [comment]
@matthew-brett  To have the Python 3.7 fix we need Cython > 0.26.1

**Comment by matthew-brett** [comment]
Can we compile with current master then?

**Comment by charris** [comment]
I think 0.26.1 is good, although there is always NumPy 1.13.4 if it isn't. I did try a build tag for the source release and it worked.
```
charris@fc [~]$ pip install --no-binary :all: --upgrade numpy --user 
Collecting numpy
  Using cached numpy-1.13.2-1.zip
Skipping bdist_wheel for numpy, due to binaries being disabled for it.
Installing collected packages: numpy
  Running setup.py install for numpy ... done
Successfully installed numpy-1.13.2
```

**Comment by matthew-brett** [comment]
Hmm - but I think what's happening there is pip thinks that this is numpy version `1.13.2-1` .  I think that's why `pip install --no-binary :all: numpy==1.13.2` fails and `pip install --no-binary :all: numpy==1.13.2-1` succeeds.

**Comment by rgommers** [comment]
On Fri, Sep 29, 2017 at 7:47 AM, Matthew Brett <notifications@github.com>
wrote:

> Can we compile with current master then?
>

Please no, we decided before not to do that (remember we tried once with
the -post0 build tag). I already see a 1.13.2-1 tarball, that looks wrong.
The whole 1.13.2 should be deleted from PyPI. Deleting only a couple of
files doesn't help, many wheels are broken and if those are removed pip
will definitely revert to source builds rather than pick up 1.13.1 wheels.

Just go for 1.13.3 straight away, that's not much more work than any other
fix.

Ralf

**Comment by rgommers** [comment]
> Can we compile with current master then?

You meant Cython master, that could be okay. Sorry, replying to issue in email client is confusing:)

Still the rest of my comment doesn't change.

**Comment by rgommers** [comment]
> (drive by comment): it's generally discouraged to delete anything from PyPI.

Eh, that doesn't apply to something that's clearly broken.

**Comment by charris** [comment]
@rgommers It was good for a bit of experimentation, which is one of the few ways to find out how PyPi actually works. I, at least, have learned something.

**Comment by matthew-brett** [comment]
Right - I meant Cython master - is it a problem compiling 1.13.3 with Cython master?

**Comment by rgommers** [comment]
> Right - I meant Cython master - is it a problem compiling 1.13.3 with Cython master?

If the numpy-wheels builds are all green, it should be okay I'd think.

**Comment by charris** [comment]
Ralf, they were all green before. There are actually a fair number of commits queued up for Cython 0.27.1, and given that 0.27 already broke NumPy in two different ways I'd rather be conservative.

**Comment by charris** [comment]
OK, 1.13.2 is gone.

**Comment by rgommers** [comment]
> I'd rather be conservative.

Yeah good idea

**Comment by charris** [comment]
Numpy 1.13.3 is out, built with Cython 0.26.1. Hopefully that will fix this problem.

**Comment by Stewori** [comment]
For the record, this issue also affected NumPy compatibility with Jython/JyNI, see https://github.com/Stewori/JyNI/issues/20. I had already found out that it was due to Cython 0.27, but was still assuming it was due to something in JyNI. Luckily, I can confirm this is fixed with NumPy 13.3!

**Comment by mattip** [comment]
Issue was  solved by removing 1.13.2 from PyPI, closing
